play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/JvmImageProcessor.java

```diff
@@ -0,0 +1,80 @@
+package com.google.wear.watchface.dfx.memory;
+
+import java.awt.image.BufferedImage;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Optional;
+import javax.imageio.ImageIO;
+import javax.imageio.stream.ImageInputStream;
+
+/** JVM-specific implementation of ImageProcessor using AWT and ImageIO. */
+public class JvmImageProcessor implements ImageProcessor {
+    @Override
+    public Optional<ImageReader> createImageReader(InputStream stream) throws IOException {
+        ImageInputStream imageInputStream = ImageIO.createImageInputStream(stream);
+        Iterator<javax.imageio.ImageReader> imageReaders =
+                ImageIO.getImageReaders(imageInputStream);
+
+        if (!imageReaders.hasNext()) {
+            return Optional.empty();
+        }
+
+        javax.imageio.ImageReader reader = imageReaders.next();
+        reader.setInput(imageInputStream);
+        return Optional.of(new JvmImageReader(reader));
+    }
+
+    private static class JvmImageReader implements ImageReader {
+        private final javax.imageio.ImageReader reader;
+
+        JvmImageReader(javax.imageio.ImageReader reader) {
+            this.reader = reader;
+        }
+
+        @Override
+        public int getWidth(int imageIndex) throws IOException {
+            return reader.getWidth(imageIndex);
+        }
+
+        @Override
+        public int getHeight(int imageIndex) throws IOException {
+            return reader.getHeight(imageIndex);
+        }
+
+        @Override
+        public int getNumImages() throws IOException {
+            // allowSearch forces the reader to return the true number of images even if the file
+            // format does not specify it, requiring an exhaustive search.
+            return reader.getNumImages(true);
+        }
+
+        @Override
+        public ImageData read(int imageIndex) throws IOException {
+            return new JvmImageData(reader.read(imageIndex));
+        }
+    }
+
+    private static class JvmImageData implements ImageData {
```

Comment:
you can return an anonymous class instead of this as well

Reply:
done